Add custom text mode

andybui01 opened the pull request:
Usage:
enter `custom` in input box and use `mod + m` to switch to custom mode.

andybui01 committed:
add custom mode

andybui01 committed:
add custom text functionality

andybui01 committed:
fix kinks in custom mode

andybui01 committed:
add auto-scroll to custom display box

andybui01 committed:
expanded punctuation mark test

andybui01 committed:
add 'custom' to list of typing mode codes

andybui01 committed:
rm debug outputs

andybui01 committed:
Merge branch 'master' of https://github.com/andybui01/typings

jakergrossman commented on main.js:
```diff
@@ -445,4 +468,15 @@ function hideThemeCenter() {
+		// Russian
+	// TODO: Add support for remaining European languages
+	// Cleaner way to do this???
+	const regex = /[\u0A00-\u0A7F\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uf900-\ufaff\uff66-\uff9f\u3131-\uD79D]+|[\u0400-\u04FFa-zA-Z0-9!:;'",.\/?!@#$%^&*()_}{\[\]-]+\'*[a-z]*/g;	
```
I think `+` and `=` should be supported. 
```suggestion
	const regex = /[\u0A00-\u0A7F\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uf900-\ufaff\uff66-\uff9f\u3131-\uD79D]+|[\u0400-\u04FFa-zA-Z0-9!:;'",.\/?!@#$%^&*()_\+\=}{\[\]-]+\'*[a-z]*/g;	

```

jakergrossman commented on index.html:
```diff
@@ -47,6 +47,9 @@ <h2 id="header">typings</h2>
             <text> / </text>
             <span id="tc-240" onclick="setTimeCount(240)">240</span>
           </span>
+          <span id="custom">
+            <span style="border-bottom: 2px solid;">CUSTOM</span>
+          </span>
```
This could be simplified
```suggestion
          <span id="custom" style="border-bottom: 2px solid;">
            CUSTOM
          </span>
```

jakergrossman commented on main.js:
```diff
@@ -31,14 +31,23 @@ function setText() {
+      const helpText = "Paste your custom text and click redo!"
+      wordList = (inputField.value === 'custom' || inputField.value === '') ? helpText.split(" "): betterSplit(inputField.value);
+      punctuation = false;
+			textDisplay.style.height = '3.2rem';
```
I think if you change `maxHeight` instead of `height`, you won't have an awkward gap under 'Paste your custom text and click redo!' but instead `textDisplay` will fit content to a MAX of 3.2 rem. This looks cleaner in my opinion.
```suggestion
			textDisplay.style.maxHeight = '3.2rem';

```

jakergrossman commented on main.js:
```diff
@@ -445,4 +468,15 @@ function hideThemeCenter() {
+	const regex = /[\u0A00-\u0A7F\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uf900-\ufaff\uff66-\uff9f\u3131-\uD79D]+|[\u0400-\u04FFa-zA-Z0-9!:;'",.\/?!@#$%^&*()_}{\[\]-]+\'*[a-z]*/g;	
+	let array = [...str.match(regex)];
+  return array;
+}
```
Make sure you have a newline at the end of your file.

briemens replied:
Please do not "reformat" the whole file! Configure your editor to use the original code style.

andybui01 replied:
Yh, I'll fix that, while changing the tab indents to spaces I must've inadvertently reformated the entire file.

jakergrossman commented on main.js:
```diff
@@ -31,14 +31,23 @@ function setText() {
   inputField.className = '';
 
   switch (typingMode) {
+		case 'custom':
```
Some loose indentation here. You are using hard tabs with a length of 4 while the source document is using soft tabs (spaces) with a length of 2. I would recommend changing the tabs in this file to 2 spaces.

andybui01 replied:
Noted. Converted all tabs to spaces.

jakergrossman commented on main.js:
```diff
@@ -111,13 +120,15 @@ function showText() {
 inputField.addEventListener('keydown', e => {
   // Add wrong class to input field
   switch (typingMode) {
+		case 'custom':
     case 'wordcount':
```
Did you write it this way so that `custom` is handled the same way as `wordcount`? It is hard to tell. Maybe something like this will convey intent better, I'm not really sure.

```suggestion
		case 'custom': case 'wordcount':
```

andybui01 replied:
Yes, I did. It's a fall-through and I was going off examples on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch).

jakergrossman replied:
Yeah it doesn’t look bad when the indentation is correct.

jakergrossman reviewed (changes requested):
Looks great! A few file formatting things (newlines and tabs) but I like it overall :+1:

andybui01 commented on main.js:
```diff
@@ -111,13 +120,15 @@ function showText() {
   }
   function inputFieldClass() {
-    if (e.key >= 'a' && e.key <= 'z' || (e.key === `'` || e.key === ',' || e.key === '.' || e.key === ';')) {
+    const puncRegex = /[!:;'",.\/?!@#$%^&*()_}{\[\]-]/;
```
```suggestion
    const puncRegex = /[!:;'",.\/?!@#$%^&*()_}{\[\]-\+\=]/;
```

Add `+` and `=` to match support for said characters in [`betterSplit()`](https://github.com/briano1905/typings/pull/68/files#r421577328).

Should the regex for punctuation marks be kept as a global constant?

jakergrossman replied:
That’s a good idea, since it should be the same for the whole app.

andybui01 committed:
fix indentation to spaces

andybui01 committed:
Apply suggestions from code review

Co-authored-by: Jake Grossman <jake.r.grossman@gmail.com>

andybui01 committed:
Merge branch 'master' of https://github.com/andybui01/typings

andybui01 committed:
add global punctuation mark regex

andybui01 commented:
I've moved the punctuation mark regex out into a global constant, as a result regex literals are no generated by concatenating strings before creating the RegExp objects themselves.

See https://github.com/briano1905/typings/pull/68/commits/d1a8ed8c1f085ed3429e009cc2a3f89330c7c28c

briemens commented:
> Usage:
> enter `custom` in input box and use `mod + m` to switch to custom mode.

Can you explain what this does? How it works? What happens when you switch to "custom mode".
Screenshot or animated gif with a demonstration would be helpful.

andybui01 commented:
> > Usage:
> > enter `custom` in input box and use `mod + m` to switch to custom mode.
> 
> Can you explain what this does? How it works? What happens when you switch to "custom mode".
> Screenshot or animated gif with a demonstration would be helpful.

When in custom mode you can paste your own text and run the typing test on said text. I'm not really familiar with making gifs but here's a screenshot of the prompt when you switch to custom mode.
![image](https://user-images.githubusercontent.com/22631610/81398380-72fe8600-916c-11ea-91f6-fa474c29ce0d.png)

SeerLite commented:
Conflicts can be resolved with `dos2unix`, on Linux at least.
I have the feeling the original repo here is badly configured though.

SeerLite commented:
Hey @andybui01 I made #92. It's based on this PR by you. I fixed the formatting with unix2dos and added some fixes myself. I also changed the functionality quite a bit so it isn't exactly the same. I hope that's fine!

andybui01 commented:
> Hey @andybui01 I made #92. It's based on this PR by you. I fixed the formatting with unix2dos and added some fixes myself. I also changed the functionality quite a bit so it isn't exactly the same. I hope that's fine!

Hey, it's fine. I've forgotten about this PR as I had to focus on uni so you're welcome to take over.
